drivers/dai/intel/ssp/ssp.c

```diff
@@ -2144,41 +2260,58 @@ static int dai_ssp_probe(struct dai_intel_ssp *dp)
 	/* allocate private data */
 	ssp = k_calloc(1, sizeof(*ssp));
 	if (!ssp) {
-		LOG_ERR("SSP%d: alloc failed", dp->index);
+		LOG_ERR("SSP%d: alloc failed", ssp_link->link_index);
 		return -ENOMEM;
 	}
 	dai_set_drvdata(dp, ssp);
 
-	ssp->state[DAI_DIR_PLAYBACK] = DAI_STATE_READY;
-	ssp->state[DAI_DIR_CAPTURE] = DAI_STATE_READY;
+	dp->state[DAI_DIR_PLAYBACK] = DAI_STATE_READY;
+	dp->state[DAI_DIR_CAPTURE] = DAI_STATE_READY;
+
+	if (ssp_link->is_power_en) {
+		return 0;
+	}
 
 #if CONFIG_INTEL_MN
 	/* Reset M/N, power-gating functions need it */
-	dai_ssp_mn_reset_bclk_divider(dp, dp->index);
+	dai_ssp_mn_reset_bclk_divider(dp, ssp_link->link_index);
 #endif
 
 	/* Enable SSP power */
-	dai_ssp_pm_runtime_en_ssp_power(dp, dp->index);
+	dai_ssp_pm_runtime_en_ssp_power(dp, ssp_link->link_index);
 
 	/* Disable dynamic clock gating before touching any register */
-	dai_ssp_pm_runtime_dis_ssp_clk_gating(dp, dp->index);
+	dai_ssp_pm_runtime_dis_ssp_clk_gating(dp, ssp_link->link_index);
+
+	ssp_link->is_power_en = true;
 
 	return 0;
 }
 
 static int dai_ssp_remove(struct dai_intel_ssp *dp)
 {
-	dai_ssp_pm_runtime_en_ssp_clk_gating(dp, dp->index);
+	struct intel_ssp_link *ssp_link = dai_get_ssp_link(dp);
+
+	if (ssp_link_is_acquired(ssp_link)) {
+		k_free(dai_get_drvdata(dp));
```

Comment:
One more, can you explain the logic here? If the link is acquired, how come we proceed to free the "'dp' object in this case?